Enable ESI caching for blog index

sergiu-popa opened the pull request:
We could update the demo app to include by default ESI caching for the blog index and the sidebar.

javiereguiluz commented:
@sergiu-popa thanks for proposing this feature. My first comment would be to please remove the output of the date in each block. It's true that it's a convenient trick to understand what's going on with ESI, but in this case it wouldn't be necessary.

Then, my question to the rest of the people that follow this repository would be: is it OK to show EASI + HttpCache in the demo app or is it too advanced?

chapay commented:
@sergiu-popa You should not update `web/config.php` as it was customized, see the conversation [here](https://github.com/symfony/symfony-demo/pull/241) and the origin [PR #81](https://github.com/symfony/symfony-demo/pull/81).

bocharsky-bw commented:
Proceeding from the fact that simple ESI implementation require minimum changes of code I think it'll be nice to have it in this demo application :+1:

bocharsky-bw commented on app/config/config.yml:
```diff
@@ -43,7 +43,7 @@ framework:
         # handler_id set to null will use default session handler from php.ini
         handler_id:  ~
-    fragments:       ~
+    fragments:       { path: /_fragment}
```
Add missing space at the end: `{ path: /_fragment }`

bocharsky-bw commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+.idea
```
This changes should be reverted. It's too personal for different users.

I suggest you to add `.idea` to the global `.gitignore` on your host machine if you always use PhpStorm for developing. Check https://help.github.com/articles/ignoring-files/#create-a-global-gitignore

voronkovich commented on app/Resources/views/base.html.twig:
```diff
@@ -105,16 +105,7 @@
-                                    </p>
-                                </div>
-                            </div>
+                            {{ render_esi(controller('AppBundle:Blog:sidebar')) }}
```
@sergiu-popa, I think you can use this (instead of creating an  action):
```twig
{{ render_esi('FrameworkBundle:Template:template', { 'defaults': { 'template': 'blog/sidebar.html.twig', 'sharedAge': 600 }}) }}
```
See
http://symfony.com/doc/current/cookbook/templating/render_without_controller.html#caching-the-static-template
http://symfony.com/doc/current/reference/twig_reference.html#render-esi

voronkovich replied:
Sorry, @sergiu-popa, it should be:
```twig
{{ render_esi(controller('FrameworkBundle:Template:template', { 'template': 'blog/sidebar.html.twig', 'sharedAge': 600 })) }}

voronkovich commented on app/config/config.yml:
```diff
@@ -26,7 +26,7 @@ framework:
     ide: sublime
 
-    # esi:             ~
+    esi:             { enabled: true }
```
@sergiu-popa, please add here a comment with a short description and a link to http://symfony.com/doc/current/book/http_cache.html#edge-side-includes

voronkovich commented on app/Resources/views/base.html.twig:
```diff
@@ -35,7 +35,7 @@
                         <div class="navbar-header">
                             <a class="navbar-brand" href="{{ path('homepage') }}">
-                                Symfony Demo
+                                Symfony Demo <small style="font-size: 10px;"><strong>ESI TIME:</strong> {{ "now"|date('H:i:s') }}</small>
```
@sergiu-popa,  I think, you could add this dates wrapped by html comments, like this:
```twig
<!-- ESI TIME: {{ "now"|date('H:i:s') }} -->
```

sergiu-popa commented:
Thank you. I removed the date output from display with HTML comments. I reverted config.php and added comment and link in config.yml. The only remaining thing is this:
`{{ render_esi('FrameworkBundle:Template:template', { 'defaults': { 'template': 'blog/sidebar.html.twig', 'sharedAge': 600 }}) }}`

I would be nice to remove the sidebarAction() and render it using the code, but I get `An exception has been thrown during the rendering of a template ("No route found for "GET Template:template"") in base.html.twig at line 108. `

Debugging using `php app/console debug:router` the route doesn't show. It should appear in the list, right? Or some routing file needs to be imported?

sergiu-popa commented:
All done. Thank you

voronkovich commented on app/config/config.yml:
```diff
@@ -43,7 +45,7 @@ framework:
         handler_id:  ~
-    fragments:       ~
+    fragments:       { path: /_fragment }
     http_method_override: true
```
@sergiu-popa, this change isn't needed. See https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L175

voronkovich commented on src/AppBundle/Controller/BlogController.php:
```diff
@@ -44,7 +44,10 @@ public function indexAction($page)
 
-        return $this->render('blog/index.html.twig', array('posts' => $posts));
+        $response = $this->render('blog/index.html.twig', array('posts' => $posts));
+        $response->setSharedMaxAge(10); // cached for 10 seconds
```
@sergiu-popa, what about to use the `@Cache(smaxage="10")` annotation here?
See http://symfony.com/doc/current/bundles/SensioFrameworkExtraBundle/annotations/cache.html#attributes

sergiu-popa replied:
Definitely more elegant with annotation, but a bit more hidden.

voronkovich replied:
@sergiu-popa, using annotations for caching is a best practice:
```
Make your controller extend the FrameworkBundle base controller and use annotations to configure routing, caching and security whenever possible.
```

See http://symfony.com/doc/current/best_practices/controllers.html

voronkovich commented on app/Resources/views/blog/sidebar.html.twig:
```diff
@@ -0,0 +1,11 @@
+<div class="section about">
```
@sergiu-popa, I would rename this template to `about.html.twig`

voronkovich commented on src/AppBundle/Controller/BlogController.php:
```diff
@@ -21,6 +21,7 @@
 use Symfony\Bundle\FrameworkBundle\Controller\Controller;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Cache;
```
@sergiu-popa, please, move this string up to other annotations import

bocharsky-bw replied:
Did you mean to sort used namespaces alphabetically?

voronkovich replied:
@bocharsky-bw, yes I did. I should look like:
```php
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Cache;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\ParamConverter;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Security;
```

voronkovich commented on app/Resources/views/base.html.twig:
```diff
@@ -105,16 +105,8 @@
-                                </div>
-                            </div>
+                            {{ render_esi(controller('FrameworkBundle:Template:template',
+                            { 'template': 'blog/about.html.twig', 'sharedAge': 600 })) }}
```
@sergiu-popa, please, add the current locale here: 
```twig
{{ render_esi(controller('FrameworkBundle:Template:template', { 
    'template': 'blog/about.html.twig',
    'sharedAge': 600,
    '_locale': app.request.locale
})) }}
```

voronkovich commented:
:+1: I hope that newcomers won't notice all this stuff :)

sergiu-popa commented:
That encouraging :+1: What about this? https://t.co/7cBQGNwGD2 A form filter that could be implemented in the admin area for many-to-many relationship, KnpPaginator compatible.

javiereguiluz commented:
@sergiu-popa it took some time, but this is definitely going to be merged. Thanks for your patience and thanks to all the reviewers for a very insightful discussion, specially to @voronkovich
